crates/transpiler/src/passes/basis_translator/mod.rs

```diff
@@ -437,176 +384,191 @@ fn extract_basis_target_circ(
-            IndexSet::from_iter(node_qarg.iter().copied());
+        let qubit_set: AhashIndexSet<Qubit> = AhashIndexSet::from_iter(node_qarg.iter().copied());
         let mut new_op: Option<OperationFromPython> = None;
         if target_basis.contains(node_obj.op.name()) || node_qarg.len() < min_qubits {
             if node_obj.op.control_flow() {
-                let OperationRef::Instruction(control_op) = node_obj.op.view() else {
-                    unreachable!("This instruction {} says it is of control flow type, but is not an Instruction instance", node_obj.op.name())
-                };
-                let mut flow_blocks = vec![];
-                let bound_obj = control_op.instruction.bind(py);
-                let blocks = bound_obj.getattr("blocks")?;
-                for block in blocks.try_iter()? {
-                    let block = block?;
-                    let dag_block: DAGCircuit = circuit_to_dag(block.extract()?, true, None, None)?;
-                    let updated_dag: DAGCircuit;
-                    (updated_dag, is_updated) = apply_translation(
-                        py,
-                        &dag_block,
-                        target_basis,
-                        instr_map,
-                        extra_inst_map,
-                        min_qubits,
-                        qargs_with_non_global_operation,
-                    )?;
-                    let flow_circ_block = if is_updated {
-                        DAG_TO_CIRCUIT
-                            .get_bound(py)
-                            .call1((updated_dag,))?
-                            .extract()?
-                    } else {
-                        block
+                Python::with_gil(|py| -> PyResult<()> {
+                    // This part is only executed through python because `ControlFlowOp`
+                    // does not exist in Rust space yet, and we need the method `replace_blocks`.
+                    // TODO: Refactor this condition block once https://github.com/Qiskit/qiskit/pull/14568 merges.
+                    let OperationRef::Instruction(control_op) = node_obj.op.view() else {
+                        unreachable!("This instruction {} says it is of control flow type, but is not an Instruction instance", node_obj.op.name())
                     };
-                    flow_blocks.push(flow_circ_block);
-                }
-                let replaced_blocks = bound_obj.call_method1("replace_blocks", (flow_blocks,))?;
-                new_op = Some(replaced_blocks.extract()?);
+                    let mut flow_blocks = vec![];
+                    let bound_obj = control_op.instruction.bind(py);
+                    let blocks = bound_obj.getattr("blocks")?;
+                    for block in blocks.try_iter()? {
+                        let block = block?;
+                        let dag_block: DAGCircuit =
+                            circuit_to_dag(block.extract()?, true, None, None)?;
```

Comment:
Even though we have the GIL, it'd be cleaner to avoid using it, right? Can we do this as:
```suggestion
                    for block in node_obj.op.block() {
                        let dag_block: DAGCircuit = circuit_to_dag(&block, true, None, None)?;
```

Reply:
For this we specifically need to end up with `QuantumCircuit` instances. And if we do this here we will end up having to convert the blocks to `QuantumCircuit` instances. This is due to `ControlFlowOp` taking in

Reply:
I updated it to use your methodology in a210bcfb9eeef47be38bc7dab418ed2f7378ee67, but this will not ever look good until #14568 merges.